src/Npgsql/NpgsqlWriteBuffer.cs

```diff
@@ -118,67 +118,90 @@ public async Task Flush(bool async)
+                    finalCt = _timeoutCts.Token;
+
+                    if (cancellationToken.CanBeCanceled)
+                    {
+                        combinedCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, _timeoutCts.Token);
+                        finalCt = combinedCts.Token;
+                    }
                 }
-                timeoutCt = _timeoutCts.Token;
-            }
 
-            try
-            {
-                if (async)
-                    await Underlying.WriteAsync(Buffer, 0, WritePosition, timeoutCt);
-                else
-                    Underlying.Write(Buffer, 0, WritePosition);
-            }
-            catch (Exception e)
-            {
-                switch (e)
+                try
                 {
-                case OperationCanceledException _:
-                // Note that mono throws SocketException with the wrong error (see #1330)
-                case IOException _ when (e.InnerException as SocketException)?.SocketErrorCode ==
-                                            (Type.GetType("Mono.Runtime") == null ? SocketError.TimedOut : SocketError.WouldBlock):
-                    Debug.Assert(e is OperationCanceledException ? async : !async);
-                    throw Connector.Break(new NpgsqlException("Exception while writing to stream", new TimeoutException("Timeout during writing attempt")));
+                    if (async)
+                        await Underlying.WriteAsync(Buffer, 0, WritePosition, finalCt);
+                    else
+                        Underlying.Write(Buffer, 0, WritePosition);
                 }
+                catch (Exception e)
+                {
+                    switch (e)
+                    {
+                    // User requested the cancellation
+                    case OperationCanceledException _ when (cancellationToken.IsCancellationRequested):
+                        throw Connector.Break(e);
+                    // Read timeout
+                    case OperationCanceledException _:
+                    // Note that mono throws SocketException with the wrong error (see #1330)
+                    case IOException _ when (e.InnerException as SocketException)?.SocketErrorCode ==
+                                                (Type.GetType("Mono.Runtime") == null ? SocketError.TimedOut : SocketError.WouldBlock):
+                        Debug.Assert(e is OperationCanceledException ? async : !async);
+                        throw Connector.Break(new NpgsqlException("Exception while writing to stream", new TimeoutException("Timeout during writing attempt")));
+                    }
 
-                throw Connector.Break(new NpgsqlException("Exception while writing to stream", e));
-            }
+                    throw Connector.Break(new NpgsqlException("Exception while writing to stream", e));
+                }
 
-            try
-            {
-                if (async)
+                try
```

Comment:
Looking at all this code duplication, we could simply unify Write and Flush and use a variable to control the exception message (that's the only reason for the duplication I think).

Reply:
Maybe we should just get rid of the different exception messages entirely? I'm not sure it's really needed and helpful in any way.

Reply:
I'm OK with that, I agree that the distinction between Write and Flush here isn't interesting.